pkg/autoscaler/autoscaler/util.go

```diff
@@ -91,11 +93,54 @@ func limitTargetReplicas(targetReplicas int32, tac *v1alpha1.TidbClusterAutoScal
 // If the minReplicas not set, the default value would be 1
 // If the Metrics not set, the default metric will be set to 80% average CPU utilization.
 // defaultTAC would default the omitted value
-func defaultTAC(tac *v1alpha1.TidbClusterAutoScaler) {
+func defaultTAC(tac *v1alpha1.TidbClusterAutoScaler, tc *v1alpha1.TidbCluster) {
 	if tac.Annotations == nil {
 		tac.Annotations = map[string]string{}
 	}
 
+	defResources := func(resources *[]v1alpha1.AutoResource, component v1alpha1.MemberType) {
+		resource := v1alpha1.AutoResource{
+			ResourceType: fmt.Sprintf("default_%s", component.String()),
+		}
+		var requests corev1.ResourceList
+		switch component {
+		case v1alpha1.TiDBMemberType:
+			requests = tc.Spec.TiDB.Requests
+		case v1alpha1.TiKVMemberType:
+			requests = tc.Spec.TiKV.Requests
+		}
+		for res, v := range requests {
+			switch res {
+			case corev1.ResourceCPU:
+				resource.CPU = v
+			case corev1.ResourceMemory:
+				resource.Memory = v
+			case corev1.ResourceStorage:
+				resource.Storage = v
+			}
+		}
+		*resources = append(*resources, resource)
+	}
+
+	if len(tac.Spec.Resources) == 0 {
+		// Construct default resource
+		if tac.Spec.TiKV != nil {
+			defResources(&tac.Spec.Resources, v1alpha1.TiKVMemberType)
+		}
+
+		if tac.Spec.TiDB != nil {
+			defResources(&tac.Spec.Resources, v1alpha1.TiDBMemberType)
+		}
+	}
+
+	defResourceTypes := func(rule *v1alpha1.AutoRule) {
+		if len(rule.ResourceTypes) == 0 {
+			for _, res := range tac.Spec.Resources {
+				rule.ResourceTypes = append(rule.ResourceTypes, res.ResourceType)
```

Comment:
Especially for the `default_tidb`, whose storage would be 0 and should not be used as the resource type for TiKV.

Reply:
what should we do if user has specified serveral resources with some having 0 storage and does not specify resource types for tikv?

Reply:
1. Should we default to `tikv.requests.storage` in TC?
2. Ignore them and use the `default_tikv`
WDYT? @howardlau1999 @rleungx 

BTW, I think we should define this field https://github.com/pingcap/tidb-operator/blob/master/pkg/apis/pingcap/v1alpha1/tidbclusterautoscaler_types.go#L114 to be a map with the key as the resource_type, otherwise, users may configure two resources with the same resource_type.

Reply:
Default to `tikv.requests.storage` is unreasonable because user may not have such resources. And we don't have `default_tikv` when user has defined some resources.

How about:
1. If user does not provide a resource with storage and specifies TiKV autoscaling, the configuration is invalid and provides an error message
2. If user provides some resources with storage and some without, we can use only those with storage when defaulting

Reply:
Change resources to `map` is OK

Reply:
@DanielZhangQD WDYT

Reply:
> Default to `tikv.requests.storage` is unreasonable because user may not have such resources. And we don't have `default_tikv` when user has defined some resources.
> 
> How about:
> 
> 1. If user provides custom resource types but does not provide a resource with storage and specifies TiKV autoscaling, the configuration is invalid and we should show an error message
> 2. If user provides some resources with storage and some without, we can use only those with storage when defaulting

OK. Then we also have to validate the storage if users have provided resource types in the rules.
cc @rleungx